piper/keystroke.py

```diff
@@ -0,0 +1,128 @@
+# Copyright (C) 2017 Jente Hidskes <hjdskes@gmail.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, write to the Free Software Foundation, Inc.,
+# 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+
+import gi
+gi.require_version("Gtk", "3.0")
+from gi.repository import Gdk, GObject, Gtk
+
+
+class KeyStroke(GObject.Object):
+    """The KeyStroke object represents a keyboard shortcut as pressed by the
+    user in the key mapping's capture mode."""
+
+    # GDK uses an offset of 8 from the keycodes defined in linux/input.h as
+    # used by evdev.
+    _XORG_KEYCODE_OFFSET = 8
+
+    def __init__(self, **kwargs):
+        """Intantiates a new KeyStroke."""
+        GObject.Object.__init__(self, **kwargs)
+        self._accel = ""
+        self._keycode = 0
+        self._keyval = 0
+        self._modifiers = []
+
+    @GObject.Property(type=str)
+    def accel(self):
+        """The accelerator of the current keystroke, as generated by
+        Gtk.accelerator_name(keyval, mask)."""
+        return self._accel
+
+    def set_key(self, keycode, keyval):
+        """Sets the given keycode and keyval as the key of this keystroke.
+
+        @param keycode The raw key code as reported by GDK, as int.
+        @param keyval The key, as int. See the `gdk/gdkkeysyms.h` header file
+                      for a complete list of GDK key codes.
+        """
+        # TODO: is a validity check necessary here? In principle, all uses of
```

Comment:
I didnt' necessarily mean verifying that the keycode belongs to the keyval, but more that both are valid integers. For example, `passing -1` here for the keycode (for whatever reason), would make `Gtk.accelerator_name(keyval, mask)` error out. So what we could do here is to check whether `keycode` and `keyval` are within a certain range, e.g. for `keyval` this could be `Gdk.KEY_BackSpace <= keyval <= Gdk.KEY_LogGrabInfo`. Though I don't think that the `keyval` values are necessarily consecutive so this might give a false sense of security, and I'm not sure what these values would be for `keycode`.